File: server/modules/git/worktree-create.ts
import { basename, join, resolve } from "node:path";
import { GitError, runGit } from "./exec.js";
import { branchExists } from "./worktree-list.js";

export interface CreateWorktreeInput {
  repoPath: string;
  branch: string;
  baseBranch?: string;
  path?: string;
  isNewBranch?: boolean;
}

export interface CreateWorktreeResult {
  path: string;
  branch: string;
  baseBranch: string | null;
}

function defaultWorktreePath(repoPath: string, branch: string): string {
  const parent = resolve(repoPath, "..");
  const repoName = basename(repoPath);
  const safeBranch = branch.replace(/\//g, "-");
  return join(parent, `${repoName}-${safeBranch}`);
}

export function createWorktree(input: CreateWorktreeInput): CreateWorktreeResult {
  const { repoPath, branch } = input;
  const exists = branchExists(repoPath, branch);
  const wantsNew = input.isNewBranch === true || (!exists && Boolean(input.baseBranch));

  if (!exists && !wantsNew && !input.baseBranch) {
    throw new GitError(
      `Branch "${branch}" does not exist. Choose a base branch to create it from.`,
      "",
    );
  }

  const targetPath = resolve(
    input.path?.trim() || defaultWorktreePath(repoPath, branch),
  );

  if (wantsNew) {
    const baseBranch = input.baseBranch?.trim();
    if (!baseBranch) {
      throw new GitError("baseBranch is required when creating a new branch", "");
    }
    if (!branchExists(repoPath, baseBranch)) {
      throw new GitError(`Base branch "${baseBranch}" does not exist`, "");
    }
    if (exists) {
      throw new GitError(`Branch "${branch}" already exists`, "");
    }
    runGit(repoPath, ["worktree", "add", "-b", branch, targetPath, baseBranch]);
    return { path: targetPath, branch, baseBranch };
  }

  if (!exists) {
    throw new GitError(`Branch "${branch}" does not exist`, "");
  }

  runGit(repoPath, ["worktree", "add", targetPath, branch]);
  return { path: targetPath, branch, baseBranch: null };
}

File: server-rs/src/git/worktree_create.rs
use super::exec::{run_git, GitError};
use super::worktree_list::branch_exists;
use std::path::{Path, PathBuf};

pub struct CreateWorktreeInput<'a> {
    pub repo_path: &'a str,
    pub branch: &'a str,
    pub base_branch: Option<&'a str>,
    pub path: Option<&'a str>,
    pub is_new_branch: Option<bool>,
}

pub struct CreateWorktreeResult {
    pub path: String,
    pub branch: String,
    pub base_branch: Option<String>,
}

fn default_worktree_path(repo_path: &str, branch: &str) -> PathBuf {
    let repo = Path::new(repo_path);
    let parent = repo.parent().unwrap_or_else(|| Path::new("."));
    let repo_name = repo
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("repo");
    let safe_branch = branch.replace('/', "-");
    parent.join(format!("{repo_name}-{safe_branch}"))
}

pub fn create_worktree(input: &CreateWorktreeInput<'_>) -> Result<CreateWorktreeResult, GitError> {
    let exists = branch_exists(input.repo_path, input.branch)?;
    let wants_new = input.is_new_branch == Some(true)
        || (!exists && input.base_branch.is_some());

    if !exists && !wants_new && input.base_branch.is_none() {
        return Err(GitError::new(
            format!(
                "Branch \"{}\" does not exist. Choose a base branch to create it from.",
                input.branch
            ),
            "",
        ));
    }

    let target_path = input
        .path
        .map(str::trim)
        .filter(|p| !p.is_empty())
        .map(PathBuf::from)
        .unwrap_or_else(|| default_worktree_path(input.repo_path, input.branch));
    let target_path = target_path
        .canonicalize()
        .unwrap_or(target_path)
        .to_string_lossy()
        .into_owned();

    if wants_new {
        let base_branch = input
            .base_branch
            .map(str::trim)
            .filter(|s| !s.is_empty())
            .ok_or_else(|| GitError::new("baseBranch is required when creating a new branch", ""))?;

        if !branch_exists(input.repo_path, base_branch)? {
            return Err(GitError::new(
                format!("Base branch \"{base_branch}\" does not exist"),
                "",
            ));
        }
        if exists {
            return Err(GitError::new(
                format!("Branch \"{}\" already exists", input.branch),
                "",
            ));
        }

        run_git(
            input.repo_path,
            &[
                "worktree",
                "add",
                "-b",
                input.branch,
                &target_path,
                base_branch,
            ],
            None,
        )?;
        return Ok(CreateWorktreeResult {
            path: target_path,
            branch: input.branch.to_string(),
            base_branch: Some(base_branch.to_string()),
        });
    }

    if !exists {
        return Err(GitError::new(
            format!("Branch \"{}\" does not exist", input.branch),
            "",
        ));
    }

    run_git(
        input.repo_path,
        &["worktree", "add", &target_path, input.branch],
        None,
    )?;
    Ok(CreateWorktreeResult {
        path: target_path,
        branch: input.branch.to_string(),
        base_branch: None,
    })
}
